### PR TITLE
Develop/improve method types

### LINKED ISSUE #87: Serialization usage unclear
Currently, serialized instances/arrays return correctly mapped objects, but they remain instances of the object.

It should be considered to make them plain JS objects instead.

### PR DESCRIPTION
**Pull request by andreas-aeschlimann**
I implemented two generics now in the deserialization methods:
- The first is for the actual data (could be an object [array] or instance [array])
- The second is for the (optional) model that we want to use for serialization

Am I right to assume that by saying `<T extends object>` we are covering the necessary use cases?

Closes #87

### CODE REVIEW

File: src/json2typescript/json-convert.ts [hunk 1]
```diff
@@ -228,7 +228,7 @@ export class JsonConvert {
      *
      * @see https://www.npmjs.com/package/json2typescript full documentation
      */
-    serialize<T>(data: any | any[], classReference?: { new(): T }): any | any[] {
+    serialize<T extends object, U extends object = {}>(data: T | T[], classReference?: { new(): U }): object | object[] {
```
Comment (andreas-aeschlimann) [inline src/json2typescript/json-convert.ts:231]: We want to  use a different generic type because we now allow that the data has actually a different structure that the class reference. I use the return type object because it should be a plain JS object now.

File: src/json2typescript/json-convert.ts [hunk 2]
```diff
@@ -338,13 +338,13 @@ export class JsonConvert {
             console.log("----------");
         }
 
-        return jsonObject;
+        return JSON.parse(JSON.stringify(jsonObject));
```
Comment (andreas-aeschlimann) [inline src/json2typescript/json-convert.ts:341]: I suggest we use this trick to make sure a plain object is returned. This actually would finally close #87.

File: src/json2typescript/json-convert.ts [hunk 3]
```diff
@@ -433,7 +433,7 @@ export class JsonConvert {
      *
      * @see https://www.npmjs.com/package/json2typescript full documentation
      */
-    deserialize<T>(json: any, classReference: { new(): T }): T | T[] {
+    deserialize<T extends object>(json: any, classReference: { new(): T }): T | T[] {
```
Comment (andreas-aeschlimann) [inline src/json2typescript/json-convert.ts:436]: Is it good to say `T extends object` or can we go even further?

**Comment by andreas-aeschlimann** [comment]
@cmolodo happy to hear your input

**Comment by tobiasschweizer** [comment]
Ok, I will check next week!

**Comment by cmolodo** [comment]
Sorry for the delay!  I think the return type from the serialize() methods needs to stay any/any[] rather than object, because otherwise if you try to modify them after serialization you get a typescript compile error, since of course object has no properties.

For example, if you add the following test to the integration tests of serialize, you'll get a compile error with the current changes:
```
it('should allow modification of the generated JSON object after serialization', () => {
  const result = jsonConvert.serialize<Cat>(cat1);
  result.catName = "New name";
});
```

File: src/json2typescript/json-convert.ts (hunk 2 above)
Comment (cmolodo) [reply src/json2typescript/json-convert.ts:341]: Is that necessary?  From the code, since it's created just as `{}`, it looks like it should be returning a plain object.  When I ran a quick test, the constructor of the result came back as an Object, not as a class constructor.  Not sure if there's some other issue I'm missing?

File: src/json2typescript/json-convert.ts (hunk 3 above)
Comment (cmolodo) [reply src/json2typescript/json-convert.ts:436]: Seems like T extends object is probably good enough?  Presumably it should only be restricted to non-primitives, but otherwise put no constraints on the type of object.

**Comment by cmolodo** [comment]
@andreas-aeschlimann Other than the serialize methods returning any rather than object, as I commented above, I think the typing changes make sense.  Thank you for fixing this, I'm sorry I didn't get to it!

File: src/json2typescript/json-convert.ts (hunk 3 above)
Comment (andreas-aeschlimann) [reply src/json2typescript/json-convert.ts:436]: > Seems like T extends object is probably good enough?  Presumably it should only be restricted to non-primitives, but otherwise put no constraints on the type of object.

I think it depends on the parameter you are passing. If I recall correcty, the serializaion sometimes keeps the original instance and just reassigns values. I will check later if that is true.

File: src/json2typescript/json-convert.ts (hunk 2 above)
Comment (andreas-aeschlimann) [reply src/json2typescript/json-convert.ts:341]: > Is that necessary?  From the code, since it's created just as `{}`, it looks like it should be returning a plain object.  When I ran a quick test, the constructor of the result came back as an Object, not as a class constructor.  Not sure if there's some other issue I'm missing?


I think it depends on the parameter you are passing. If I recall correcty, the serializaion sometimes keeps the original instance and just reassigns values. I will check later if that is true.

File: src/json2typescript/json-convert.ts (hunk 1 above)
Comment (tobiasschweizer) [reply src/json2typescript/json-convert.ts:231]: I don't understand `U extends object = {}`. Doesn't this mix a type and an instance?

I guess what you want to say is that U is of type object (a subclass), but why create an empty object? That should be handled by the implementation of the method.

Comment (tobiasschweizer) [reply src/json2typescript/json-convert.ts:231]: or is this like a type annotation for the return type of an arrow function?

Comment (tobiasschweizer) [reply src/json2typescript/json-convert.ts:231]: see https://www.typescriptlang.org/docs/handbook/generics.html#generic-types

Comment (andreas-aeschlimann) [reply src/json2typescript/json-convert.ts:231]: The second parameter of type U is optional. Afaik this is the way to let the compiler know that U belongs to an optional parameter and must not be passed (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults):

Without default generic value, you could not write `serialize<User>(user)`, you would have to write `serialize<User, User>(user)`, even though the second generic is useless.

**Comment by andreas-aeschlimann** [comment]
> Sorry for the delay! I think the return type from the serialize() methods needs to stay any/any[] rather than object, because otherwise if you try to modify them after serialization you get a typescript compile error, since of course object has no properties.
> 
> For example, if you add the following test to the integration tests of serialize, you'll get a compile error with the current changes:
> 
> ```
> it('should allow modification of the generated JSON object after serialization', () => {
>   const result = jsonConvert.serialize<Cat>(cat1);
>   result.catName = "New name";
> });
> ```

Does this not work unless the type is any? Is the object type immutable after initialization?

File: src/json2typescript/json-convert.ts (hunk 1 above)
Comment (tobiasschweizer) [reply src/json2typescript/json-convert.ts:231]: I think in general, if the type can be inferred, you do not have to pass the type also if there is no default:

> `let output = identity("myString");  // type of output will be 'string'`
Notice that we didn’t have to explicitly pass the type in the angle brackets (<>); the compiler just looked at the value "myString", and set T to its type.  While type argument inference can be a helpful tool to keep code shorter and more readable, you may need to explicitly pass in the type arguments as we did in the previous example when the compiler fails to infer the type, as may happen in more complex examples.
https://www.typescriptlang.org/docs/handbook/generics.html#hello-world-of-generics

But in this case, I think the compiler cannot figure it out because it is the return type.
So the following applies:
> If a default type is specified and inference cannot choose a candidate, the default type is inferred.
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults

In short, I think what you do is correct. :-)

**Comment by cmolodo** [comment]
Yeah - the object type is explicitly defined as having only certain properties.  It's just like any other variable with a declared type or class, you get a Typescript error if you try to add or access a property that isn't defined on that type/class (even if they actually exist on the object).

It's also similar to creating an empty object without explicitly setting the variable type - you won't be able to set any properties on it unless you do.
```
const obj = {};
// This will generate a Typescript error
obj.propertyName = "New property";

const anyObj: any = {};
// By setting type explicitly to any (or another class/type), now you can access/add properties
anyObj.propertyName = "New property";
```
Since in this case we don't want the type to be a specific class but just a plain JS object, we have to use type any as the return signature.

File: src/json2typescript/json-convert.ts (hunk 2 above)
Comment (andreas-aeschlimann) [reply src/json2typescript/json-convert.ts]: I think you are right, it was unnecessary.

**Comment by andreas-aeschlimann** [comment]
> Yeah - the object type is explicitly defined as having only certain properties. It's just like any other variable with a declared type or class, you get a Typescript error if you try to add or access a property that isn't defined on that type/class (even if they actually exist on the object).
> 
> It's also similar to creating an empty object without explicitly setting the variable type - you won't be able to set any properties on it unless you do.
> 
> ```
> const obj = {};
> // This will generate a Typescript error
> obj.propertyName = "New property";
> 
> const anyObj: any = {};
> // By setting type explicitly to any (or another class/type), now you can access/add properties
> anyObj.propertyName = "New property";
> ```
> 
> Since in this case we don't want the type to be a specific class but just a plain JS object, we have to use type any as the return signature.

This is fixed now and I will merge it.
